text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,230 @@
+
+The feature suggested here is unlikely to be used by "organizations" as this would put independent concerns in the same Rust API.  By "organizations", I mean a group of people who are coming together to build likely related crates, under the same "brand", likely developed in multiple repos under a GitHub organization.
+
+
+The motivation here is distinct from the general problem of squatting -- with general squatting, someone else might come up with a cool crate name before you do. However, with `projectname-foo` crates, it's more of a case of third parties "muscling in" on a name you have already chosen and are using.
+
+# Guide-level explanation
+
+The owners of the `foo` crate may provide other crates under the `foo` namespace, like `foo::bar`.  For users, this makes its official status clearer and makes it easier to discover.
+
+Users import these crates in Cargo.toml as normal:
+
+```toml
+[dependencies]
+"foo" = "1.0.42"
+"foo::bar" = "3.1"
+```
+
+They will then access this through a facade made of `foo` and all `foo::*` crates, for example:
+
+```rs
+let baz = foo::bar::Baz::new();
+foo::render(baz);
+```
+
+Some reasons for `foo`s owner to consider using namespaces:
+- Avoid name conflicts with third-party authors (since they are reserved)
+- Improve discoverability of official crates
+- As an alternative to feature flags for optional subsystems
+- When different parts of your API might have different compatibility guarantees
+
+When considering this, keep in mind:
+- Does it makes sense for this new crate to be presented in the `foo` facade?
+- How likely is a crate to move into or out of the namespace?
+  - Moving the crate in or out of a namespace is a breaking change though it can be worked around by having the old crate re-export the new crate but that does add extra friction to the process.
+  - There is not currently a mechanism to raise awareness with users that a crate has migrated into or out of a namespace and you might end up leaving users behind.
+- If users import both `foo` and `foo::bar` but `foo` also has a `bar` item in its API that isn't just `foo::bar` re-exported, then rustc will error.
+
+Only the owners of `foo` may _create_ the `foo::bar` crate (and all owners of `foo` are implicitly owners of `foo::bar`). After the `foo::bar` crate is created, additional per-crate publishers may be added who will be able to publish subsequent versions as usual.
+
+# Reference-level explanation
+
+`::` is now considered valid inside crate names on Crates.io. For now, we will restrict crate names to having a single `::` in them, not at the beginning or end of the name, but this can be changed in the future.
```

Comment:
If we still support implicit features being created on behalf of namespaced dependencies, then this will also make `::` valid for feature names.

Should we support `::` in features (`foo::bar`) or require the user to namespace these (`dep:foo::bar`)?

Reply:
I'd go with required namespacing for now